juju/juju
# Model info fix: related to list models failures.

## Description of change

Under some circumstances, list models could fail. This was particularly apparent in the cases where models were marked for destruction or have been in process of being imported. The listing itself was not failing. However, the code that retrieved model information for display could not handle partial information.
Previous code expected all information for a model to be present and accessible.
This PR changes the behavior:
* For models that are alive and active, we do expect to receive all information. When we cannot, we consider it a fatal error and surface it;
* For models that are not alive - dead or dying, i.e. the models that are in some phases of lifecycle leading to their total destruction - or the models that are being imported, we do not expect to receive all information. When we encounter NotFound errors, they are not considered fatal. This allows us to send partial information or, at least, all available-at-the-time-of-the-request information. Note that all other types of errors are still considered fatal.

The most interesting part of this PR is in apiserver/modelmanager/modelmanager.go.
The rest of the PR is dealing with the ability to display partial information without too many surprises. It also contains a comprehensive amount of testing to ensure that most common combinations of partial information are formatted in a manner that is pleasing to the eye and is not-surprising for automatic digestion.

## QA steps

Since this change is for the models in particular phases in lifecycle, there is a bit of timing competition with Juju to test this change reliably. This is most notoriously better in CI or on some very slow machines. When manually verifying, you would need to be able to catch a model in either dying state or when migrating, in importing phase. On my machine, mongo powers through and does the right things quickly, so it was very hard to catch.
Steps to re-produce:
1. bootstrap
2. add model with various entities - machines and users worked bets for me;
3. destroy model while listing models
You should  observe any errors.

In CI,  model migration related tests were the best at detecting the issue intermittently.

## Documentation changes

This PR changes output of list models as well as show models, essentially omitting fields with empty values.
For list models, in tabular output, empty values are shown as  '-' to ensure that lines are easily deduced for larger outputs.
If documentation has sample outputs for these two commands, they may need to be updated.

## Bug reference
https://bugs.launchpad.net/juju/+bug/1674805
For https://bugs.launchpad.net/juju/+bug/1676279, this PR addresses the list models failure only. The bug itself about canceling model destruction still stands.
